File: passgen-app/passgen-web/src/main/java/com/passgen/core/actors/SupervisorActor.java
package com.passgen.core.actors;

import akka.actor.*;
import akka.event.Logging;
import akka.event.LoggingAdapter;
import scala.concurrent.duration.Duration;

import java.util.concurrent.TimeUnit;

public class SupervisorActor extends UntypedActor {
    private final LoggingAdapter LOG = Logging.getLogger(getContext().system(), this);

    private final ActorRef marshaller;
    private final ActorRef generator;
    private final ActorRef fin;

    public SupervisorActor() {
        fin = getContext().actorOf(Props.create(FinActor.class), "fin");
        generator = getContext().actorOf(Props.create(GeneratePasswordActor.class, fin), "generator");
        marshaller = getContext().actorOf(Props.create(MarshallerActor.class, generator), "marshaller");
    }

    public SupervisorStrategy supervisorStrategy() {
        return new AllForOneStrategy(2, Duration.apply(5, TimeUnit.SECONDS), (t) -> {
            if (t instanceof NumberFormatException) {
                LOG.info("Restart actors, error: {}", t);
                return SupervisorStrategy.restart();
            } else {
                LOG.info("Escalate, error: {}", t);
                return SupervisorStrategy.escalate();
            }
        });
    }

    @Override
    public void onReceive(Object m) throws Exception {
        LOG.info("Forward message: {}", m);
        marshaller.forward(m, getContext());
    }
}

File: passgen-app/passgen-scala/src/main/scala/com/passgen/core/actors/SupervisorActor.scala
package com.passgen.core.actors

import java.util.concurrent.TimeUnit

import akka.actor.SupervisorStrategy.{Escalate, Restart}
import akka.actor._
import akka.event.slf4j.SLF4JLogging

import scala.concurrent.duration.Duration

class SupervisorActor extends Actor with SLF4JLogging {
  val fin: ActorRef = context.actorOf(Props.create(classOf[FinActor]), "fin")
  val generator: ActorRef = context.actorOf(Props.create(classOf[GeneratePasswordActor], fin), "generator")
  val marshaller: ActorRef = context.actorOf(Props.create(classOf[MarshallerActor], generator), "marshaller")

  def receive = {
    case m =>
      log.debug("Forward message: {}", m)
      marshaller forward m
  }

  override def supervisorStrategy = AllForOneStrategy(2, Duration.apply(5, TimeUnit.SECONDS)) {
    case e: NumberFormatException =>
      log.info("Restart actors, error: {}", e)
      Restart
    case other =>
      log.info("Escalate, error: {}", other)
      Escalate
  }
}